src/main/java/org/jvnet/hudson/test/WarExploder.java

```diff
@@ -74,18 +78,31 @@ private static File explode() throws Exception {
             if(new File(d,".jenkins").exists()) {
                 File dir = new File(d,"war/target/jenkins");
                 if(dir.exists()) {
-                    System.out.println("Using jenkins.war resources from "+dir);
+                    LOGGER.log(Level.INFO, "Using jenkins.war resources from {0}", dir);
                     return dir;
                 }
             }
         }
 
         // locate jenkins.war
+        File war;
         URL winstone = WarExploder.class.getResource("/winstone.jar");
-        if(winstone==null)
-            // impossible, since the test harness pulls in jenkins.war
-            throw new AssertionError("jenkins.war is not in the classpath. If you are running this from the core workspace, run 'mvn install' to create the war image in war/target/jenkins. If running from a plugin, try using `mvn clean test` (cf. JENKINS-45245).");
-        File war = Which.jarFile(Class.forName("executable.Executable"));
+        if (winstone != null) {
+            war = Which.jarFile(Class.forName("executable.Executable"));
+        } else {
+            // JENKINS-45245: work around incorrect test classpath in IDEA. Note that this will not correctly handle timestamped snapshots; in that case use `mvn test`.
```

Comment:
Unclear what this means. Will a `jenkins.version` snapshot dependency break with this? If so, is there a way to detect it here and output this comment as an explicit error message (perhaps as part of the assertion error on 99?)

Reply:
It just means that given a `jenkins-core.version` of `2.99-20180101.1234` or whatever and `jenkins-war.version` `2.99-20180101.5678`, from the perspective of `WarExploder` we only know that we are running out of `…/jenkins-core/2.99-SNAPSHOT/jenkins-core-2.99-SNAPSHOT.jar` and thus that we should load `…/jenkins-war/2.99-SNAPSHOT/jenkins-war-2.99-SNAPSHOT.war`. It does not necessarily know the specific version to use.

Basically, a locally built snapshot dep should be fine, but do not expect a checkout of a downstream plugin PR using an upstream core PR to allow you to run tests from IDEA without using `mvn test`. Which you would rarely want to do anyway, since if you were running tests locally, it is probably because you are testing local changes including to core.

Reply:
Of course all of this is speculative—I cannot myself try it.

Reply:
Perhaps `if (version.contains("SNAPSHOT")) { /* log something saying we might pick up the wrong war *? }`? But this can go ahead without that, could always be added later.

Reply:
Yeah I would leave this to an IDEA user actually trying to use snapshot core deps to further test & refine.

Reply:
Fine for me.